<a href="https://colab.research.google.com/github/richierichijkl/Projects/blob/main/OptionPricing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
from sklearn.preprocessing import StandardScaler

# Load your option pricing data (Assuming data is in CSV format)
option_data = pd.read_csv("/content/drive/MyDrive/Datasets/tsla_2019_2022.csv")

# Display the first few rows to understand the structure
option_data.head()






<ipython-input-3-4012a68cb689>:5: DtypeWarning: Columns (8,9,10,11,12,15,17,18,20,21,23,24,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  option_data = pd.read_csv("/content/drive/MyDrive/Datasets/tsla_2019_2022.csv")


[QUOTE_UNIXTIME]   [QUOTE_READTIME]  [QUOTE_DATE]   [QUOTE_TIME_HOURS]  \
0        1556740800   2019-05-01 16:00    2019-05-01                 16.0   
1        1556740800   2019-05-01 16:00    2019-05-01                 16.0   
2        1556740800   2019-05-01 16:00    2019-05-01                 16.0   
3        1556740800   2019-05-01 16:00    2019-05-01                 16.0   
4        1556740800   2019-05-01 16:00    2019-05-01                 16.0   

    [UNDERLYING_LAST]  [EXPIRE_DATE]   [EXPIRE_UNIX]   [DTE]  [C_DELTA]  \
0              233.98     2019-05-03      1556913600     2.0   0.984650   
1              233.98     2019-05-03      1556913600     2.0   0.983710   
2              233.98     2019-05-03      1556913600     2.0   0.984580   
3              233.98     2019-05-03      1556913600     2.0   0.991870   
4              233.98     2019-05-03      1556913600     2.0   0.993410   

   [C_GAMMA]  ...   [P_LAST]   [P_DELTA]  [P_GAMMA]   [P_VEGA]   [P_THETA]  \
0   0.000550  ...   0.010000   -0.000710   0.000030   0.000460   -0.009750   
1   0.000670  ...   0.020000   -0.001090   0.000090   0.000580   -0.010100   
2   0.000690  ...   0.020000   -0.001220   0.000120   0.001130   -0.014350   
3   0.000490  ...   0.010000   -0.001340   0.000090   0.001390   -0.014650   
4   0.000390  ...   0.010000   -0.001760   0.000080   0.001050   -0.014670   

      [P_RHO]     [P_IV]   [P_VOLUME]  [STRIKE_DISTANCE]  \
0    0.000000   2.225480   147.000000              104.0   
1   -0.000010   2.083490    12.000000               99.0   
2    0.000000   2.023590    15.000000               94.0   
3    0.000000   1.895040     0.000000               89.0   
4    0.000000   1.768120    91.000000               84.0   

    [STRIKE_DISTANCE_PCT]  
0                   0.444  
1                   0.423  
2                   0.402  
3                   0.380  
4                   0.359  

[5 rows x 33 columns]

In [ ]:
option_data.shape

(2659259, 33)

In [ ]:
# Clean the data (e.g., handle missing values, remove outliers, etc.)
option_data.dropna(inplace=True)  # Drop rows with missing values

In [ ]:
option_data.shape

(2659259, 33)

In [ ]:
option_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2659259 entries, 0 to 2659258
Data columns (total 33 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   [QUOTE_UNIXTIME]        int64  
 1    [QUOTE_READTIME]       object 
 2    [QUOTE_DATE]           object 
 3    [QUOTE_TIME_HOURS]     float64
 4    [UNDERLYING_LAST]      float64
 5    [EXPIRE_DATE]          object 
 6    [EXPIRE_UNIX]          int64  
 7    [DTE]                  float64
 8    [C_DELTA]              object 
 9    [C_GAMMA]              object 
 10   [C_VEGA]               object 
 11   [C_THETA]              object 
 12   [C_RHO]                object 
 13   [C_IV]                 object 
 14   [C_VOLUME]             object 
 15   [C_LAST]               object 
 16   [C_SIZE]               object 
 17   [C_BID]                object 
 18   [C_ASK]                object 
 19   [STRIKE]               float64
 20   [P_BID]                object 
 21   [P_ASK]                object 

In [ ]:
# Print all column names to inspect them
print(option_data.columns)


Index(['[QUOTE_UNIXTIME]', ' [QUOTE_READTIME]', ' [QUOTE_DATE]',
       ' [QUOTE_TIME_HOURS]', ' [UNDERLYING_LAST]', ' [EXPIRE_DATE]',
       ' [EXPIRE_UNIX]', ' [DTE]', ' [C_DELTA]', ' [C_GAMMA]', ' [C_VEGA]',
       ' [C_THETA]', ' [C_RHO]', ' [C_IV]', ' [C_VOLUME]', ' [C_LAST]',
       ' [C_SIZE]', ' [C_BID]', ' [C_ASK]', ' [STRIKE]', ' [P_BID]',
       ' [P_ASK]', ' [P_SIZE]', ' [P_LAST]', ' [P_DELTA]', ' [P_GAMMA]',
       ' [P_VEGA]', ' [P_THETA]', ' [P_RHO]', ' [P_IV]', ' [P_VOLUME]',
       ' [STRIKE_DISTANCE]', ' [STRIKE_DISTANCE_PCT]'],
      dtype='object')


In [ ]:
# Strip spaces from column names
option_data.columns = option_data.columns.str.strip()
option_data.columns = option_data.columns.str.replace(r'[\[\]]', '', regex=True)
# Check again
print(option_data.columns)


Index(['QUOTE_UNIXTIME', 'QUOTE_READTIME', 'QUOTE_DATE', 'QUOTE_TIME_HOURS',
       'UNDERLYING_LAST', 'EXPIRE_DATE', 'EXPIRE_UNIX', 'DTE', 'C_DELTA',
       'C_GAMMA', 'C_VEGA', 'C_THETA', 'C_RHO', 'C_IV', 'C_VOLUME', 'C_LAST',
       'C_SIZE', 'C_BID', 'C_ASK', 'STRIKE', 'P_BID', 'P_ASK', 'P_SIZE',
       'P_LAST', 'P_DELTA', 'P_GAMMA', 'P_VEGA', 'P_THETA', 'P_RHO', 'P_IV',
       'P_VOLUME', 'STRIKE_DISTANCE', 'STRIKE_DISTANCE_PCT'],
      dtype='object')


In [ ]:
# Convert date columns to datetime format
option_data['QUOTE_DATE'] = pd.to_datetime(option_data['QUOTE_DATE'])
option_data['EXPIRE_DATE'] = pd.to_datetime(option_data['EXPIRE_DATE'])

# Convert columns like 'C_DELTA', 'C_GAMMA' etc., to numeric, forcing errors to NaN
numeric_columns = ['C_DELTA', 'C_GAMMA', 'C_VEGA', 'C_THETA', 'C_RHO', 'C_IV',
                   'C_VOLUME', 'C_LAST', 'C_BID', 'C_ASK', 'P_BID', 'P_ASK',
                   'P_SIZE', 'P_LAST', 'P_DELTA', 'P_GAMMA', 'P_VEGA', 'P_THETA',
                   'P_RHO', 'P_IV', 'P_VOLUME']

In [ ]:
# Convert the columns to numeric, coercing errors to NaN
for col in numeric_columns:
    option_data[col] = pd.to_numeric(option_data[col], errors='coerce')

In [ ]:
option_data.shape

(2659259, 33)

In [ ]:
# Calculate the 'days to expiration' feature
option_data['days_to_expiration'] = (option_data['EXPIRE_DATE'] - pd.to_datetime('today')).dt.days

In [ ]:
option_data.shape

(2659259, 34)

In [ ]:
# Normalize numerical features (e.g., 'STRIKE', 'UNDERLYING_LAST', 'STRIKE_DISTANCE')
scaler = StandardScaler()
numerical_features = ['STRIKE', 'UNDERLYING_LAST', 'STRIKE_DISTANCE', 'STRIKE_DISTANCE_PCT',
                      'C_IV', 'P_IV']  # List of features that need normalization

In [ ]:
option_data.shape

(2659259, 34)

In [ ]:
# Standardize the numerical features
option_data[numerical_features] = scaler.fit_transform(option_data[numerical_features])

# Final cleaned and preprocessed option data
print(option_data.head())

   QUOTE_UNIXTIME     QUOTE_READTIME QUOTE_DATE  QUOTE_TIME_HOURS  \
0      1556740800   2019-05-01 16:00 2019-05-01              16.0   
1      1556740800   2019-05-01 16:00 2019-05-01              16.0   
2      1556740800   2019-05-01 16:00 2019-05-01              16.0   
3      1556740800   2019-05-01 16:00 2019-05-01              16.0   
4      1556740800   2019-05-01 16:00 2019-05-01              16.0   

   UNDERLYING_LAST EXPIRE_DATE  EXPIRE_UNIX  DTE  C_DELTA  C_GAMMA  ...  \
0        -1.320954  2019-05-03   1556913600  2.0  0.98465  0.00055  ...   
1        -1.320954  2019-05-03   1556913600  2.0  0.98371  0.00067  ...   
2        -1.320954  2019-05-03   1556913600  2.0  0.98458  0.00069  ...   
3        -1.320954  2019-05-03   1556913600  2.0  0.99187  0.00049  ...   
4        -1.320954  2019-05-03   1556913600  2.0  0.99341  0.00039  ...   

   P_DELTA  P_GAMMA   P_VEGA  P_THETA    P_RHO      P_IV P_VOLUME  \
0 -0.00071  0.00003  0.00046 -0.00975  0.00000  1.359523    147.0

In [ ]:
# Fill NaN values with the mean of the column
option_data['C_LAST'].fillna(option_data['C_LAST'].mean(), inplace=True)


<ipython-input-18-0d9bb6166f92>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  option_data['C_LAST'].fillna(option_data['C_LAST'].mean(), inplace=True)


In [ ]:
option_data.shape

(2659259, 34)

In [ ]:
'''from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Assuming your target variable is 'C_LAST' (call option price) or 'P_LAST' (put option price)
# X will be all features, and y will be the target variable
X = option_data[['STRIKE', 'UNDERLYING_LAST', 'days_to_expiration', 'C_IV', 'P_IV']]
y = option_data['C_LAST']  # Target: Option price

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the model (Random Forest Regressor)
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Mean Absolute Error: {mae}")
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")'''


KeyboardInterrupt: 

In [ ]:
# Use a smaller subset for quick testing
option_data_small = option_data.sample(frac=0.1, random_state=42)

X_small = option_data_small[['STRIKE', 'UNDERLYING_LAST', 'days_to_expiration', 'C_IV', 'P_IV']]
y_small = option_data_small['C_LAST']

X_train, X_test, y_train, y_test = train_test_split(X_small, y_small, test_size=0.2, random_state=42)

model.fit(X_train, y_train)


RandomForestRegressor(random_state=42)

In [ ]:

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Mean Absolute Error: {mae}")
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")

Mean Absolute Error: 53.752631653152854
Mean Squared Error: 12233.89417114097
R-squared: 0.8020914862022589


Your model's performance is fairly strong, but let’s break it down:

Interpretation of Your Results
Mean Absolute Error (MAE):

53.75 means that, on average, your model's predictions are off by $53.75 per option price.
Depending on the typical price of options in your dataset, this could be acceptable or too high.
If the average option price is in the hundreds, this might be a reasonable error, but if options are usually in the tens, this is too high.
Mean Squared Error (MSE):

12,233.89 is the squared error, which amplifies larger mistakes.
This is useful for penalizing large prediction mistakes, but it’s hard to interpret directly.
R-squared (R²):

0.802 (80.2%) means that your model explains about 80.2% of the variance in the option prices.
This is a strong result—anything above 0.75 is generally considered a good fit.
However, there's still 19.8% of variation that your model doesn’t capture.
Is This a Good Model?
R² of 0.80 is solid, meaning the model captures most of the trends in the data.
MAE of 53.75 is acceptable if option prices are high (e.g., in the $500+ range). If options are much cheaper, this might be too large an error.
MSE is high but mostly useful for comparing models.
How to Improve the Model?
If you want to improve accuracy and reduce error:

Feature Engineering

Include volatility indicators, macro data, or moving averages.
Add features like intraday price trends, open interest, or Greeks interactions.
Use a More Advanced Model

Try Gradient Boosting Models (XGBoost, LightGBM, CatBoost), which often outperform Random Forest on structured data.

In [ ]:
pip install xgboost
